architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+        pass
+```
+
+At the top level we propose to have this `class Dataloader` which is an abstract class 
+and requires functions to process the data config proposed above.
+
+We also propose a full length config verification code which preceeds the call to function 
+`Dataloader.process_data_config` as the function expects a `DataConfig` object.
+
+The data loader needs to support custom data handlers which are provided by users of the library
+or even predefined handlers which need to be registered with the data loader class using the 
+call `DataLoader.register_data_handler`.
+
+The reason to have a top level dataloader class is to ensure we have separate classes for
+hugging face and stateful data loader (fms-fsdp implementation).
+This is also` selected from the argument in the config `dataloader.type`
+
+## How are handlers provided and registered - 
+
+Data handlers are python callables which can be called on single/few samples of data and can perform
+things like applying chat template, tokenising the data, applying tools like jinja template or even
+things like encoding or decoding multi modal formats like images/audio for processing by the model.
+
+The abstract data loader class provides a way to register datahandler against a `name` which is a string.
+The data handler config `DataHandlerConfig` taken by `execute_data_handlers` represents a DAG of data handling
+routines which are to be executed on the data. 
+
+For standard HF API you can think of these as the HF Processing routines. Which could be Map/Filter/Select operations
+We implement most of the routines as map and because of this even the tokenisation of data which is done today
+in fms-hf-tuning via `tuning/utils/preprocessing_utils.py::get_preprocessed_dataset` can be retained as a data 
+handler which performs tokenization.
+
+The implementation is flexible enough for users to specify their own implementation of data handling routines
+which can be called by the data loader as part of its execution.
+
+To this end, one way to design is we can provide the users and API on like the one shown in the `Dataloader` class
+which they can utilise to register custom data handlers, in this case however the user needs to use `fms-hf-tuning` as
+a module but not via the implementation of its `main` functionality.
+
+Please note that our implementation needs to support certain predefined built-in handlers like `apply_chat_template`
+or `tokenize` which user can request just by a name.
+
+For example see this implementation - https://github.ibm.com/ai4code-wisdom/platform/blob/main/modelops/modelops/train.py#L251
+
+# Implementation of HuggingFace Data Loader.
+
+HF Dataloader or a dataloader which used Hugging Face API is implemented as an instance of the `Dataloader` class.
+
+When the dataloader goes through each `DataSetConfig`
+
+The HF dataloader implements functionality to process different type of files via its `load_dataset` factory.
```

Comment:
It is not clear to me fi the plan is to implement a new `DataLoader` abstraction or what. I feel taht from a data standpoint, the abstraction interface should be the `DataSet` not the `DataLoader`. THis is because in `accelerate`, the It has its own `DataLoader` implementation. You dont want to touch that.

Reply:
https://github.com/huggingface/transformers/issues/31441

Reply:
This plan is not to superimpose on the data loader implementations of accelerate or others rather just to say that this internal abstraction will use HF/Accelerate API while other can implement data loaders which are experimental but not yet available in HF. 

Thanks for the pointer on the `StatefulDataloader` we can definitely evaluate that for our consideration but the idea to have a design like this is to allow flexibility of implementing custom dataloaders like the one in our `fms-fsdp` repo inside `fms-hf-tuning`.

Reply:
I believe accelerate stance on the data loader is that it shouldn’t be user implemented. I feel you can cover most use cases by customising dataset, data collator, sampler or batch sampler. Do you have a use case where these are not sufficient

Reply:
use case of custom dataloader is also not clear to me. Seems like from use cases you defined so far, we just need
1. preprocessing functions for selected data types like parquet,chat etc
2. code for sampling and thus mixing different formats of datasets in one training 

They do not warrant a new dataLoader. Also to mention a big reason of using standard HF APIs in this repo was simplicity and ease of maintenance. With the direction being - any new logic should be contributed to HF , so we can leverage standard APIs.

Reply:
This section needs to be removed as agreed - no customization of DataLoaders themselves